src/link/MachO.zig

```diff
@@ -4386,25 +4368,43 @@ fn allocateTextBlock(self: *MachO, text_block: *TextBlock, new_block_size: u64,
     return vaddr;
 }
 
-pub fn addExternFn(self: *MachO, name: []const u8) !SymbolWithLoc {
-    log.debug("adding new extern function '{s}' with dylib ordinal 1", .{name});
+pub fn addExternFn(self: *MachO, name: []const u8) !u32 {
+    const sym_name = try std.fmt.allocPrint(self.base.allocator, "_{s}", .{name});
+    const already_defined = self.symbol_resolver.contains(sym_name);
+
+    if (already_defined) {
+        const resolv = self.symbol_resolver.get(sym_name) orelse unreachable;
+        self.base.allocator.free(sym_name);
+        return resolv.where_index;
+    }
+
+    log.debug("adding new extern function '{s}' with dylib ordinal 1", .{sym_name});
     const import_sym_index = @intCast(u32, self.imports.items.len);
     try self.imports.append(self.base.allocator, .{
-        .n_strx = try self.makeString(name),
+        .n_strx = try self.makeString(sym_name),
         .n_type = macho.N_UNDF | macho.N_EXT,
         .n_sect = 0,
         .n_desc = packDylibOrdinal(1),
         .n_value = 0,
     });
-    const resolv = .{
+    try self.symbol_resolver.putNoClobber(self.base.allocator, sym_name, .{
         .where = .import,
         .where_index = import_sym_index,
-    };
-    try self.symbol_resolver.putNoClobber(self.base.allocator, try self.base.allocator.dupe(u8, name), resolv);
+    });
+
     const stubs_index = @intCast(u32, self.stubs.items.len);
     try self.stubs.append(self.base.allocator, import_sym_index);
     try self.stubs_map.putNoClobber(self.base.allocator, import_sym_index, stubs_index);
-    return resolv;
+
+    // TODO discuss this. The caller context expects codegen.InnerError{ OutOfMemory, CodegenFail },
+    // which obviously doesn't include file writing op errors. So instead of trying to write the stub
+    // entry right here and now, queue it up and dispose of when updating decl.
```

Comment:
The linker is supposed to handle file writing errors, and add a compile error to Module in the appropriate place (there are several places that track compile errors) and then return `error.CodegenFail`.

Reply:
Perfect! Exactly as I imagined it in my head!